File: rate2015.pas
PROGRAM CALCOLO_INTERESSI_E_RATE_(input,output);
uses crt,printer;
  VAR  tipo_calcolo,tipo_rata,tipo_stampa:CHAR;
       v,k,j,h,conta,conta2,conta3,operazioni,rate:INTEGER;
       m,l,MINIMO,MASSIMO,rata_semestrale,percentuale,e,x,prestito,inter:REAL;

  PROCEDURE Do_Loop1(zz:integer);
    BEGIN
      clrscr;
      conta:=1;
      inter:=x*percentuale/v;
      x:=(x+inter-e);
      writeln('  ',conta:7,'  ',e:19:2,'  ',x:20:2,'  ',inter:19:2);
      REPEAT
        conta:=conta+1;inter:=x*percentuale/v;x:=(x+inter-e);
        IF zz=1 THEN BEGIN
             writeln('  ',conta:7,'  ',e:19:2,'  ',x:20:2,'  ',inter:19:2);
             IF (conta=24) or (conta=48) or (conta=72) THEN BEGIN
                 readln;clrscr;
               END;
             IF conta=rate THEN BEGIN
                  writeln;
                  IF tipo_rata='m' THEN writeln('   numero rate    rata mensile       capitale residuo          interesse')
                                 ELSE writeln('   numero rate   rata semestrale     capitale residuo          interesse');
                END
           END
      UNTIL conta=rate;
      IF zz=1 THEN BEGIN
           gotoxy(35,25);write('premi invio ');gotoxy(30,35);readln;
         END;
    END;

  PROCEDURE  Do_Loop2(zz:integer);
    BEGIN
      clrscr;
      x:=prestito;conta:=1;
      x:=x+(x*e/v)-rata_semestrale;clrscr;
      writeln('  ',conta:10,'  ',e:28:9,'  ',x:26:2);
      REPEAT
        conta:=conta+1;x:=x+(x*e/v)-rata_semestrale;
        IF zz=1 THEN BEGIN
             writeln('  ',conta:10,'  ',e:28:9,'  ',x:26:2);
             IF (conta=24) or (conta=48) or (conta=72) THEN BEGIN
                  readln;clrscr;
                END;
             IF conta=rate THEN BEGIN
                  writeln;
                  writeln('      numero rate         percentuale interesse       capitale residuo');
                END
           END
      UNTIL conta=rate;
      IF zz=1 THEN BEGIN
           gotoxy(35,25);write('premi invio ');gotoxy(30,35);readln;
         END;
  END;

  PROCEDURE Calcolo_Interessi;FORWARD;
  PROCEDURE Calcolo_Rate;FORWARD;

  PROCEDURE Scelta_Tipo_Rate;
    BEGIN
      REPEAT
        clrscr;
        gotoxy(25,4);writeln('SCELTA RATE MENSILI O SEMESTRALI');
        gotoxy(30,9);writeln('M    PER RATA MENSILE');
        gotoxy(30,13);writeln('S    PER RATA SEMESTRALE');
        gotoxy(30,17);writeln('X    PER FINE LAVORO');
        gotoxy(36,22);writeln('SCELTA:');
        gotoxy(45,22);readln(tipo_rata);
        CASE tipo_rata OF
                    'm': BEGIN v:=1200; k:=9000; j:=8; END;
                    's': BEGIN v:=200; k:=1500; j:=8; END;
                    END;
      UNTIL (tipo_rata='m') or (tipo_rata='s') or (tipo_rata='x');
    END;

  PROCEDURE Menu1;
    BEGIN
      operazioni:=0; 
      clrscr;
      gotoxy(29,4);write('R    PER CALCOLO RATE');
      gotoxy(29,8);write('I    PER CALCOLO INTERESSI');
      gotoxy(29,12);write('S    PER SCELTA RATE');
      gotoxy(29,16);write('X    PER FINE LAVORO');
      gotoxy(27,22);write('PREMERE      R   I   S   X ');
    END;

  PROCEDURE Principale;
    BEGIN
      gotoxy(55,22);readln(tipo_calcolo);
      CASE tipo_calcolo OF
                    'r': Calcolo_Rate;
                    'i': Calcolo_Interessi;
                    's': Scelta_Tipo_Rate;
                    'x': BEGIN END
                  ELSE
                    Menu1;
      END; (* fine CASE OF *)
    END;

  PROCEDURE Stampa_Loop1(zz:integer); FORWARD;

  PROCEDURE Calcolo_Rate;
    BEGIN
      clrscr;gotoxy(26,5);write('PRESTITO LIRE: ');
      gotoxy(26,10);write('PERCENTUALE: ');
      gotoxy(26,15);write('NUMERO RATE: ');
      gotoxy(21,22);write('STAMPA A VIDEO O SU CARTA (v/c): ');
      gotoxy(41,5);readln(prestito);
      gotoxy(39,10);readln(percentuale);
      gotoxy(39,15);readln(rate);
      gotoxy(54,22);readln(tipo_stampa);
      IF (rate>=2) AND (rate<=32000) THEN
        IF percentuale*rate<k  THEN
          IF (prestito>=MINIMO) AND (prestito<=MASSIMO) THEN BEGIN
            e:=prestito/rate;conta3:=0;m:=10000000000.0;
            REPEAT
              operazioni:=operazioni+1;x:=prestito;
              Do_Loop1(0);
              IF x>0 THEN e:=e+m;
                IF x<0 THEN BEGIN
                           e:=e-m;conta3:=conta3+1;
                                     CASE conta3 OF
                                       1: m:=10000000000.0;
                                       2: m:=1000000000;
                                       3: m:=100000000;
                                       4: m:=10000000;
                                       5: m:=1000000;
                                       6: m:=100000;
                                       7: m:=10000;
                                       8: m:=1000;
                                       9: m:=100;
                                      10: m:=10;
                                      11: m:=1;
                                      12: m:=0.1;
                                      13: m:=0.01;
                                     END;
                         END;
            UNTIL (conta3=14) OR (operazioni=250);
            e:=e+0.001;x:=prestito;
            IF tipo_stampa='c' THEN Stampa_Loop1(1)
                               ELSE Do_Loop1(1);
          END
          ELSE BEGIN (* valore del prestito errato *)
            gotoxy(10,24);write('A PRESTITO METTERE UN NUMERO DA  4  A  12  CIFRE:  PREMI INVIO ');
            readln;
          END
        ELSE BEGIN (* percentuale*rate>=k *)
          gotoxy(16,24);write('PERCENTUALE O RATE TROPPO GRANDI:  PREMI INVIO ');
          readln;
        END
      ELSE BEGIN (* numero rate errato *)
        gotoxy(13,24);write('IMMETTERE ALMENO 2 RATE O MENO DI 32000:  PREMI INVIO ');
        readln;
      END
    END;

  PROCEDURE Stampa_Loop2; FORWARD;

  PROCEDURE Calcolo_Interessi;
    BEGIN
      clrscr;gotoxy(26,5);writeln('PRESTITO LIRE: ');
      gotoxy(26,10);IF tipo_rata='m' THEN writeln('RATA MENSILE:')
                                   ELSE writeln('RATA SEMESTRALE:');
      gotoxy(26,15);writeln('NUMERO RATE:');
      gotoxy(21,22);writeln('STAMPA A VIDEO O SU CARTA (v/c): ');
      gotoxy(41,5);readln(prestito);
      IF tipo_rata='m' THEN gotoxy(40,10)
                       ELSE gotoxy(43,10);
      readln(rata_semestrale);
      gotoxy(39,15);readln(rate);
      gotoxy(54,22);readln(tipo_stampa);
      IF rata_semestrale*rate<prestito*j THEN
        IF rate*rata_semestrale>=prestito THEN
          IF (prestito>=MINIMO) AND (prestito<=MASSIMO) THEN BEGIN
            conta2:=0;e:=1;l:=10;
              REPEAT
                operazioni:=operazioni+1;
                Do_Loop2(0);
                IF x<0 THEN e:=e+l;
                IF x>0 THEN BEGIN
                           e:=e-l;conta2:=conta2+1;
                                     CASE conta2 OF
                                       1: l:=1;
                                       2: l:=0.1;
                                       3: l:=0.01;
                                       4: l:=0.001;
                                       5: l:=0.0001;
                                       6: l:=0.00001;
                                       7: l:=0.000001;
                                       8: l:=0.0000001;
                                       9: l:=0.00000001;
                                       10: l:=0.000000001;
                                     END;
                         END;
              UNTIL (conta2=11) OR (operazioni=470);e:=e+0.0000000001;
              IF tipo_stampa='v' THEN Do_Loop2(1)
                                 ELSE Stampa_Loop2;
          END
            ELSE BEGIN
              gotoxy(10,24);write('A PRESTITO METTERE UN NUMERO DA  4  A  12  CIFRE:  PREMI INVIO ');
              readln;
            END
          ELSE BEGIN
            gotoxy(20,24);write('OPERAZIONE IMPOSSIBILE:  PREMI INVIO ');
            readln;
          END
        ELSE BEGIN
          gotoxy(15,24);write('PERCENTUALE O RATE TROPPO GRANDI:  PREMI INVIO ');
          readln;
        END
    END;

  PROCEDURE Stampa_Loop1;
    BEGIN
      writeln(lst); writeln(lst);
      writeln(lst,'        PRESTITO LIRE: ',prestito:18:0);
      writeln(lst,'        INTERESSE: ',percentuale:22:9);
      writeln(lst,'        NUMERO RATE: ',rate:20);
      writeln(lst); writeln(lst);
      conta:=1;clrscr;inter:=x*percentuale/v;x:=(x+inter-e);
      writeln(lst,'  ',conta:7,'  ',e:19:2,'  ',x:19:2,'  ',inter:18:2);
      REPEAT
        conta:=conta+1;inter:=x*percentuale/v;x:=(x+inter-e);
        writeln(lst,'  ',conta:7,'  ',e:19:2,'  ',x:19:2,'  ',inter:18:2);
        IF conta=rate
        THEN BEGIN
            writeln(lst);
            IF tipo_rata='m' THEN
                               writeln(lst,'   numero rate   rata mensile        capitale residuo         interesse')
                             ELSE
                               writeln(lst,'   numero rate  rata semestrale      capitale residuo         interesse');
        END;
      UNTIL conta=rate;
      gotoxy(35,25);write('premi invio ');gotoxy(30,35);readln;
    END;

  PROCEDURE Stampa_Loop2;
    BEGIN
      writeln(lst); writeln(lst);
      writeln(lst,'        PRESTITO LIRE: ',prestito:18:0);
      IF tipo_rata='m'
             THEN writeln(lst,'        RATA MENSILE: ',rata_semestrale:19:0)
             ELSE writeln(lst,'        RATA SEMESTRALE: ',rata_semestrale:16:0);
      writeln(lst,'        NUMERO RATE: ',rate:20);
      writeln(lst); writeln(lst);
      x:=prestito; conta:=1;
      x:=x+(x*e/v)-rata_semestrale;
      writeln(lst,'  ',conta:7,'  ',e:26:9,'  ',x:26:2);
      REPEAT
        conta:=conta+1; x:=x+(x*e/v)-rata_semestrale; 
        writeln(lst,'  ',conta:7,'  ',e:26:9,'  ',x:26:2);
        IF conta=rate THEN BEGIN
                          writeln(lst);
                          writeln(lst,'   numero rate      percentuale interesse          capitale residuo');
                        END
      UNTIL conta=rate;
      gotoxy(35,25);write('premi invio ');gotoxy(30,35);readln;
    END;

BEGIN
  (* imposto valore costanti *)
  MINIMO:=1000; MASSIMO:=1000000000000.0;

  tipo_calcolo:='0';
  textbackground(blue);textcolor(yellow);
  Scelta_Tipo_Rate;
  REPEAT
    IF tipo_calcolo<>'x' THEN
                     BEGIN
                       Menu1;
                       IF tipo_rata<>'x' THEN Principale;
                     END;
  UNTIL (tipo_calcolo='x') or (tipo_rata='x');
END.

File: rate2015.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import gi
gi.require_version('Gtk', '3.0')
gi.require_version('PangoCairo', '1.0')
from gi.repository import Gtk, Gio, GLib, Pango, PangoCairo
import sys # per controllo versione python
import math

class Handler:
    k = 9000 ; v = 1200 # valori iniziali per rata mensile
    J = 8 # valore COSTANTE per calcolo interessi

    def onDeleteMainWindow(self, *args):
        Gtk.main_quit(*args)

    def onDeleteOutputWindow (self, outputWin, event):
        print ( outputWin.get_size() )
        outputWin.hide()
        return True

    def rataMensile(self, radiobutton):
        if radiobutton.get_active():
            self.k = 9000
            self.v = 1200
            builder.get_object( "lbl_valore_rata" ).set_text( "Valore rata mensile" )

    def rataSemestrale(self, radiobutton):
        if radiobutton.get_active():
            self.k = 1500
            self.v = 200
            builder.get_object( "lbl_valore_rata" ).set_text( "Valore rata semestrale" )

    def calcolaRate(self, button):

        def Do_Loop1(zz):
            x = prestito
            conta = 1
            inter = x * percentuale / self.v
            x = x + inter - e
            if zz == 1: # memorizzo il calcolo della prima rata
                output = "\n"
                output += "  {0:7}   {1:19.2f}   {2:20.2f}   {3:19.2f}" \
                    .format(conta, e, x, inter)
                output += "\n"
            while True:
                conta += 1
                inter = x * percentuale / self.v
                x = x + inter - e
                if zz == 1: # memorizzo i calcoli relativi a rate 2..n
                    output += "  {0:7}   {1:19.2f}   {2:20.2f}   {3:19.2f}" \
                        .format(conta, e, x, inter)
                    output += "\n"
                if conta == nro_rate :
                    break
            if zz == 1: # visualizzo risultati in finestra modale
                output += "\n"
                if builder.get_object( "rata mensile" ).get_active():
                    output += ("    numero rate        rata mensile         capitale residuo          interesse    ")
                else:
                    output += ("    numero rate       rata semestrale       capitale residuo          interesse    ")
                output += "\n"

                txt_buffer = builder.get_object( "output_buffer" )
                txt_buffer.set_text(output)

                outputWin.resize(1050,510)
                outputWin.show_all()
            return x

        print( "Calcolo Rate in corso..." )

        percentuale = builder.get_object( "percentuale" ).get_value()
        nro_rate = builder.get_object( "nro_rate" ).get_value_as_int()

        # controllo correttezza valori inseriti
        if percentuale * nro_rate > self.k :
            # errore!
            dialog = builder.get_object( "error_percentuale_dialog" )
            dialog.run()

            dialog.hide()
            return

        # effettuo il calcolo
        prestito = builder.get_object( "prestito" ).get_value_as_int()
        e = prestito/nro_rate
        operazioni = 0
        conta3 = 0
        m = 10000000000.0
        lista_m = (         None,       # 0   non verrà mai usato
                   10000000000.0,       # 1
                      1000000000,       # 2
                       100000000,       # 3
                        10000000,       # 4
                         1000000,       # 5
                          100000,       # 6
                           10000,       # 7
                            1000,       # 8
                             100,       # 9
                              10,       # 10
                               1,       # 11
                               0.1,     # 12
                               0.01,    # 13
                               0.01)    # 14  mantengo valore del 13
        while True:
            # ricerchiamo il corretto valore di "e"
            operazioni += 1
            x = Do_Loop1(0)
            if x > 0 :
                e = e + m
            elif x < 0 :
                e = e - m
                conta3 += 1
                m = lista_m[conta3]
            if (conta3 == 14) or (operazioni == 250) :
                break
        e += 0.001
        x = prestito
        # ora conosciamo il valore di "e", rifacciamo il calcolo e stampiamolo
        Do_Loop1(1)
        print( "Fine Calcolo Rate" )

    def calcolaInteressi(self, button):

        def Do_Loop2(zz):
            x = prestito
            conta = 1
            x = x + (x*e/self.v) - valore_rata
            if zz == 1: # stampa calcoli relativi a rata 1
                output = "\n"
                output += "   {0:7}  {1:19.2f}    {2:20.2f}".format(conta, e, x)
                output += "\n"

            while True:
                conta += 1
                x = x + (x*e/self.v) - valore_rata
                if zz == 1 : # stampa calcoli relativi a rate 2..n
                    output += "   {0:7}  {1:19.2f}    {2:20.2f}".format(conta, e, x)
                    output += "\n"
                if conta == nro_rate :
                    break
            if zz == 1: # stampa piè di pagina
                output += "\n"
                output += ("    numero rate    percentuale interesse    capitale residuo    ")
                output += "\n"

                txt_buffer = builder.get_object( "output_buffer" )
                txt_buffer.set_text(output)

                outputWin.resize(815,510)
                outputWin.show_all()
            return x

        print( "Calcolo Interessi in corso...\n" )

        prestito = builder.get_object( "prestito" ).get_value()
        valore_rata = builder.get_object( "valore_rata" ).get_value()
        nro_rate = builder.get_object( "nro_rate" ).get_value()

        # controllo correttezza valori inseriti:
        if (valore_rata * nro_rate >= prestito * self.J) or (valore_rata * nro_rate <  prestito) :
            # errore!
            dialog = builder.get_object( "error_percentuale_dialog" )
            dialog.run()

            dialog.hide()
            return
        # effettuo il calcolo
        operazioni = 0
        conta2 = 0
        e = 1
        l = 10
        lista_l = (         None,   # 0   non verrà mai usato
                               1,   # 1
                             0.1,   # 2
                            0.01,   # 3
                           0.001,   # 4
                          0.0001,   # 5
                         0.00001,   # 6
                        0.000001,   # 7
                       0.0000001,   # 8
                      0.00000001,   # 9
                     0.000000001,   # 10
                     0.000000001    # 11  mantengo valore del 10
                  )

        while True:
            # ricerchiamo il corretto valore di "e"
            operazioni += 1
            x = Do_Loop2(0)
            if x < 0:
                e = e+l
            elif x >0:
                e = e-l
                conta2 += 1
                l = lista_l[conta2]
            if conta2 == 11 or operazioni == 470:
                break
        e = e + 0.0000000001
        # ora conosciamo il valore di "e" rifacciamo il calcolo e stampiamolo
        Do_Loop2(1)
        print( "Fine Calcolo Interessi" )

    def pageChanged (self, stack, event):
        if stack.get_visible_child_name() == "pagina rate":
            btn_calcola.set_label( "Calcola Rate" )
            btn_calcola.disconnect_by_func(self.calcolaInteressi)
            btn_calcola.connect( "clicked", self.calcolaRate )
        else:
            btn_calcola.set_label( "Calcola Interessi" )
            btn_calcola.disconnect_by_func(self.calcolaRate)
            btn_calcola.connect( "clicked", self.calcolaInteressi)

    ######----------               STAMPA               ----------######

    def stampa (self, button):
        HEADER_HEIGHT = 10 * 72 / 25.4
        HEADER_GAP = 3 * 72 / 25.4

        o_buffer = builder.get_object( "output_buffer" )
        start, end = o_buffer.get_bounds()
        output = o_buffer.get_text(start, end, True)
        parent = builder.get_object( "outputWindow" )
        tipo_calcolo = btn_calcola.get_label()
        print ("stampo")

        def prepara_stampa():
            self.operation = Gtk.PrintOperation()
            print_data = {
                          'filename': output,
                          'font_size': 12.0,
                          'lines_per_page': 0,
                          'lines': None,
                          'num_lines': 0,
                          'num_pages': 0
                         }

            self.operation.connect('begin-print', begin_print, print_data)
            self.operation.connect('draw-page', draw_page, print_data)
            self.operation.connect('end-print', end_print, print_data)

            self.operation.set_use_full_page(False)
            self.operation.set_unit(Gtk.Unit.POINTS)
            self.operation.set_embed_page_setup(True)

            settings = Gtk.PrintSettings()

            dir = GLib.get_user_special_dir(GLib.UserDirectory.DIRECTORY_DOCUMENTS)
            if dir is None:
                dir = GLib.get_home_dir()
            if settings.get(Gtk.PRINT_SETTINGS_OUTPUT_FILE_FORMAT) == 'ps':
                ext = '.ps'
            elif settings.get(Gtk.PRINT_SETTINGS_OUTPUT_FILE_FORMAT) == 'svg':
                ext = '.svg'
            else:
                ext = '.pdf'

            uri = "file://%s/gtk-demo%s" % (dir, ext)
            settings.set(Gtk.PRINT_SETTINGS_OUTPUT_URI, uri)
            self.operation.set_print_settings(settings)

        def begin_print(operation, print_ctx, print_data):
            height = print_ctx.get_height() - HEADER_HEIGHT - HEADER_GAP
            print_data['lines_per_page'] = \
                math.floor(height / print_data['font_size'])

            s = output

            print_data['lines'] = s.split('\n')
            print_data['num_lines'] = len(print_data['lines'])

            print_data['num_pages'] = \
                (print_data['num_lines'] - 1) / print_data['lines_per_page'] + 1

            operation.set_n_pages(print_data['num_pages'])

        def draw_page(operation, print_ctx, page_num, print_data):
            cr = print_ctx.get_cairo_context()
            width = print_ctx.get_width()

            cr.rectangle(0, 0, width, HEADER_HEIGHT)
            cr.set_source_rgb(0.8, 0.8, 0.8)
            cr.fill_preserve()

            cr.set_source_rgb(0, 0, 0)
            cr.set_line_width(1)
            cr.stroke()

            layout = print_ctx.create_pango_layout()
            desc = Pango.FontDescription('sans 14')
            layout.set_font_description(desc)

            layout.set_text(tipo_calcolo, -1)
            (text_width, text_height) = layout.get_pixel_size()

            if text_width > width:
                layout.set_width(width)
                layout.set_ellipsize(Pango.EllipsizeMode.START)
                (text_width, text_height) = layout.get_pixel_size()

            cr.move_to((width - text_width) / 2,
                       (HEADER_HEIGHT - text_height) / 2)
            PangoCairo.show_layout(cr, layout)

            page_str = "%d/%d" % (page_num + 1, print_data['num_pages'])
            layout.set_text(page_str, -1)

            layout.set_width(-1)
            (text_width, text_height) = layout.get_pixel_size()
            cr.move_to(width - text_width - 4,
                       (HEADER_HEIGHT - text_height) / 2)
            PangoCairo.show_layout(cr, layout)

            layout = print_ctx.create_pango_layout()

            desc = Pango.FontDescription('monospace')
            desc.set_size(print_data['font_size'] * Pango.SCALE)
            layout.set_font_description(desc)

            cr.move_to(0, HEADER_HEIGHT + HEADER_GAP)
            lines_pp = int(print_data['lines_per_page'])
            num_lines = print_data['num_lines']
            data_lines = print_data['lines']
            font_size = print_data['font_size']
            line = page_num * lines_pp

            for i in range(lines_pp):
                if line >= num_lines:
                    break

                layout.set_text(data_lines[line], -1)
                PangoCairo.show_layout(cr, layout)
                cr.rel_move_to(0, font_size)
                line += 1

        def end_print(operation, print_ctx, print_data):
            pass

        prepara_stampa()
        a = Gtk.PrintOperationAction.PRINT_DIALOG
        result = self.operation.run(a, parent)

        if result == Gtk.PrintOperationResult.ERROR:
            print("Errore")
            message = self.operation.get_error()

            dialog = Gtk.MessageDialog(parent,
                                       0,
                                       Gtk.MessageType.ERROR,
                                       Gtk.ButtonsType.CLOSE,
                                       message)

            dialog.run()
            dialog.destroy()

        output = builder.get_object( "output_buffer" )



######----------                MAIN                ----------######


# controllo che versione python sia la 3.x
if sys.version[0] == "2":
    # errore! eseguito programma con python 2.x
    print( "\a" ) # ding!
    print( "ERRORE! rilevato python 2" )
    print( "Eseguire il programma con la versione 3.x di Python" )
    print( "" )
    raise SystemExit

GUI = './finestraPrincipale.ui'

builder = Gtk.Builder()
builder.add_from_file( GUI )
btn_calcola = builder.get_object( "calcola" )
builder.connect_signals( Handler() )


window = builder.get_object( "mainWindow" )
outputWin = builder.get_object( "outputWindow" )
window.show_all()

Gtk.main()
